tracer/src/Datadog.Trace/SpanContextPropagator.cs

```diff
@@ -184,28 +146,26 @@ public SpanContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> ge
                 // Tag name is normalized during Tracer instantiation so use as-is
                 if (!string.IsNullOrWhiteSpace(headerNameToTagName.Value))
                 {
-                    yield return new KeyValuePair<string, string>(headerNameToTagName.Value, headerValue);
+                    yield return new KeyValuePair<string, string?>(headerNameToTagName.Value!, headerValue);
```

Comment:
Is `ExtractHeaderTags` being used in a hot path? If the answer is yes maybe we can replace the `yield return` with a custom enumerator to avoid allocations.

Reply:
> used in a hot path?

Only is users configure `DD_TRACE_HEADER_TAGS` to add header values as span tags.

In this PR I was focusing on the `Inject`/`Extract` methods which are used for trace propagation. I can deal with `ExtractHeaderTags` in a separate PR. Maybe move it to another helper class since it's unrelated to propagation.